Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Hole_filling/Triangulate_hole_polyline.h

```diff
@@ -1194,7 +1204,250 @@ class Triangulate_hole_polyline {
+  maxv = (CGAL::max)(maxv, x);
+  maxv = (CGAL::max)(maxv, y);
+  maxv = (CGAL::max)(maxv, z);
+  if (maxv <= FT(0)) {
+    return false; // a good plane does not exist for sure!
+  }
+
+  const Plane_3 plane = Plane_3(centroid, avg_normal);
+  FT avg_sq_distance = FT(0);
+  for (std::size_t i = 0; i < n; ++i) {
+    const Point_3& p = points[i];
+    const Point_3  q = projection_3(plane, p);
+    avg_sq_distance += squared_distance_3(p, q);
+  }
+  avg_sq_distance /= static_cast<FT>(n);
+  // std::cout << "avg squared distance: " << avg_sq_distance << std::endl;
+
+  if (avg_sq_distance > max_sq_distance) {
+    return false; // the user distance criteria are not satisfied!
+  }
+
+  // std::cout << "The hole seems to be near planar." << std::endl;
+  return true;
+}
+
+template <
+  typename PointRange, // need size()
+  typename Tracer,
+  typename Validity_checker,
+  typename Traits
+>
+bool
+triangulate_hole_polyline_with_cdt(const PointRange& points,
+                                   Tracer& tracer,
+                                   const Validity_checker& is_valid,
+                                   const Traits& traits)
+{
+  typedef typename Traits::FT FT;
+  typedef typename Traits::Point_3 Point_3;
+  typedef typename Traits::Vector_3 Vector_3;
+  typedef typename Traits::Collinear_3 Collinear_3;
+
+  // Compute an average normal of the hole.
+  const Collinear_3 collinear_3 =
+    traits.collinear_3_object();
+
+  std::vector<Point_3> P(boost::begin(points), boost::end(points));
+  CGAL_assertion(P.size() >= 3);
+  if (P.front() != P.back()) {
+    P.push_back(P.front());
+  }
+
+  FT x = FT(0), y = FT(0), z = FT(0);
+  std::size_t num_normals = 0;
+  const Point_3& ref_point = P[0];
+  const std::size_t size = P.size() - 1;
+  for (std::size_t i = 1; i < size - 1; ++i) {
+    const std::size_t ip = i + 1;
+
+    const Point_3& p1 = ref_point; // 3 points, which form a triangle
+    const Point_3& p2 = P[i];
+    const Point_3& p3 = P[ip];
+
+    // Skip in case we have collinear points.
+    if (collinear_3(p1, p2, p3)) {
+      continue;
+    }
+
+    // Computing the normal of a triangle.
+    const Vector_3 n = CGAL::normal(p1, p2, p3);
+    // If it is a positive normal ->
+    if (
+      ( n.x() >  FT(0)                                    ) ||
+      ( n.x() == FT(0) && n.y() >  FT(0)                  ) ||
+      ( n.x() == FT(0) && n.y() == FT(0) && n.z() > FT(0) )) {
+      x += n.x(); y += n.y(); z += n.z();
+    } else { // otherwise invert ->
+      x -= n.x(); y -= n.y(); z -= n.z();
+    }
+    ++num_normals;
+  }
+
+  if (num_normals < 1) {
+    return false;
+  }
+
+  // Setting the final normal.
+  x /= static_cast<FT>(num_normals);
+  y /= static_cast<FT>(num_normals);
+  z /= static_cast<FT>(num_normals);
+  const Vector_3 avg_normal = Vector_3(x, y, z);
+  // std::cout << "avg normal: " << avg_normal << std::endl;
+
+  // Checking the hole planarity.
+  const FT max_sq_distance = FT(4) / FT(10000);
```

Comment:
Yep. I did not change it because we have not yet decided in the discussion above if we should publish the planarity function on its own. If we do not have a clear path now, I can simply put it in the named parameters and document.

Reply:
Ok, done.